Accept contributions! 💝 (#12)
opendp/tumult-analytics: CONTRIBUTING.md

# Contributing

We are happy to accept external contributions! 💖

## Contact

First, let us know what you would like to contribute. Feel free to:

- report a bug or request a feature on [GitHub](https://github.com/opendp/tumult-analytics/issues);
- send general queries to info@opendp.org, or email security@opendp.org if it is related to security;
- ask any question on our [Slack][slack] instance.

[slack]: https://join.slack.com/t/opendp/shared_invite/zt-1aca9bm7k-hG7olKz6CiGm8htI2lxE8w

## Local development

### Installation

> We recommend using a Python virtual environment to do local development, for example using [pyenv](https://github.com/pyenv/pyenv): it is useful to make sure you're working with a supported Python version, and minimize interactions with your system Python environment.

Reviewer: ```suggestion
We recommend using the lowest supported python version for local development. Tools like [pyenv](https://github.com/pyenv/pyenv) may be useful for managing your python version.
```

pyenv doesn't actually manage virtual environments, despite the name. It just managed installed python versions. Poetry handles creating the virtual environments.

At some point (soon?) we probably want to switch these instructions to recommend `uv` instead (cc @tmager ), but I don't see a super easy way to substitute `uv` for `pyenv`.

> We use [Poetry](https://python-poetry.org/) for dependency management during development. To set up your environment, start by installing the prerequisites listed in the [installation instructions](https://docs.tmlt.dev/platform/latest/installation.html#prerequisites), then install Poetry using `pip install poetry`, and finally install our dev dependencies by running `poetry install` from the root of this repository.

Author: Done!

### Basic usage

You can then locally run our linters and tests by running:
```bash
make lint
make test
```
from the repository root directory. There are a few other commands available in the [Makefile](./Makefile) which can come in handy. 

Note that some operating systems, including macOS, include versions of `make` that are too old to run this project's [Makefile](./Makefile) correctly. macOS users should [install a newer version of make using Homebrew](https://formulae.brew.sh/formula/make#default).

> Behind the scenes, these commands use the Poetry environment, and rely on [nox](https://nox.thea.codes/en/stable/index.html) for test automation. For more information about how it works, see our [nox-utils](https://github.com/opendp/tumult-tools/tree/main/nox-utils) repository.

Reviewer: ```suggestion
Behind the scenes, these commands use the Poetry environment, and rely on [nox](https://nox.thea.codes/en/stable/index.html) for test automation. You can get a bit more fine-grained control and access additional tools by running nox commands directly (see [this tutorial](https://nox.thea.codes/en/stable/tutorial.html)). For a list of available nox sessions, you can run `poetry run nox --list`.
```
